speaker/bgp_controller.go

```diff
@@ -150,61 +158,167 @@ func (c *bgpController) ShouldAnnounce(l log.Logger, name string, svc *v1.Servic
 	return ""
 }
 
-// Called when either the peer list or node labels have changed,
+// Called when either the peer list or node annotations/labels have changed,
 // implying that the set of running BGP sessions may need tweaking.
 func (c *bgpController) syncPeers(l log.Logger) error {
-	var (
-		errs          int
-		needUpdateAds bool
-	)
+	var totalErrs int
+	var needUpdateAds int
+
+	// Update peer BGP sessions.
+	update, errs := c.syncBGPSessions(l, c.peers)
+	needUpdateAds += update
+	totalErrs += errs
+	l.Log("op", "syncBGPSessions", "needUpdate", update, "errs", errs, "msg", "done syncing peer BGP sessions")
+
+	// Update node peer BGP session.
+	if c.nodePeer != nil {
+		update, errs = c.syncBGPSessions(l, []*peer{c.nodePeer})
+		needUpdateAds += update
+		totalErrs += errs
+		l.Log("op", "syncBGPSessions", "needUpdate", update, "errs", errs, "msg", "done syncing node peer BGP session")
+	}
+
+	if needUpdateAds > 0 {
+		// Some new sessions came up, resync advertisement state.
+		if err := c.updateAds(); err != nil {
+			l.Log("op", "updateAds", "error", err, "msg", "failed to update BGP advertisements")
+			return err
+		}
+	}
+	if totalErrs > 0 {
+		return fmt.Errorf("%d BGP sessions failed to start", errs)
+	}
+	return nil
+}
+
+// Attempt to create a BGP peer from node annotations and/or labels if peer
+// autodiscovery is configured.
+func (c *bgpController) discoverNodePeer(l log.Logger) {
+	if c.peerAutodiscovery == nil {
+		if c.nodePeer != nil {
+			l.Log("op", "discoverNodePeer", "msg", "peer autodiscovery not configured - removing node peer")
+			c.deleteNodePeer(l)
+		}
+		return
+	}
+
+	// Parse (or re-parse) node peer configuration from the Node object.
+	discovered, err := parseNodePeer(l, c.peerAutodiscovery, c.nodeAnnotations, c.nodeLabels)
+	if err != nil {
+		// Log an error without returning to let the user know why peer
+		// autodiscovery failed for this node. We continue execution
+		// because we still want to remove any outdated node peer which may
+		// exist.
+		l.Log("op", "discoverNodePeer", "error", err, "msg", "peer autodiscovery failed")
+	}
+	nodePeerExists := c.nodePeer != nil
+
+	if discovered == nil {
```

Comment:
Initially I didn't want to combine the two cases above into one case because doing so would make it an "error" when no node peer is discovered whereas there are cases where I don't think of this case as an error, for example when the user deliberately disables autodiscovery for a node using node selectors. However, I was able to refactor things by moving the "should we discover" check from `parseNodePeer()` into `DiscoverNodePeer()`, which makes more sense to me anyway. I've also edited the logs a little bit and I think things make sense now. Following this refactoring, `parseNodePeer()` returns either a peer or an error, and in case we don't need to attempt discovery, `parseNodePeer()` isn't even called.

Thanks for the feedback.

Reply:
Thanks!